File: OOUP/lab3-eclipse/src/main/java/hr/fer/ouup/model/plugins/Parrot.java
package hr.fer.ouup.model.plugins;

import hr.fer.ouup.model.Animal;

public class Parrot extends Animal {

	private String animalName;

	public Parrot(String name) {
		this.animalName = name;
	}

	@Override
	public String name() {
		return animalName;
	}

	@Override
	public String greet() {
		return "Dobar dan dragi korisniče!";
	}

	@Override
	public String menu() {
		return "sjemenke";
	}

}

File: OOUP/lab3/prvi/parrot.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef char const* (*PTRFUN)();

struct Animal{
  PTRFUN* vtable;
  char const* name;

  // vtable entries:
  // 0: char const* name(void* this);
  // 1: char const* greet();
  // 2: char const* menu();
};

char const* name(struct Animal* this){
    return this->name;
}

char const* greet(){
    return "Dobar dan dragi korisnice!";
}

char const* menu(){
    return "sjemenke";
}
PTRFUN parrotTable[3] = {name, greet, menu};

void construct(struct Animal *animal, char const* animalName){
    animal->vtable = parrotTable;
    animal->name = animalName;
}

struct Animal* create(char const* animalName){
    struct Animal *memoryParrot = (struct Animal*) malloc(sizeof(struct Animal*));
    construct(memoryParrot, animalName);
    return memoryParrot;
} 